### PR TITLE
Add shims for the zlib functions used in System.IO.Compression on Unix.

### PR DESCRIPTION
**Pull request by eerhardt**
@ianhays @sokket @nguerrera @stephentoub

### CODE REVIEW

File: src/Native/System.IO.Compression.Native/pal_zlib.cpp [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+static_assert(PAL_Z_BESTCOMPRESSION == Z_BEST_COMPRESSION, "");
+static_assert(PAL_Z_DEFAULTCOMPRESSION == Z_DEFAULT_COMPRESSION, "");
+
+static_assert(PAL_Z_FILTERED == Z_FILTERED, "");
+static_assert(PAL_Z_HUFFMANONLY == Z_HUFFMAN_ONLY, "");
+static_assert(PAL_Z_RLE == Z_RLE, "");
+static_assert(PAL_Z_FIXED == Z_FIXED, "");
+static_assert(PAL_Z_DEFAULTSTRATEGY == Z_DEFAULT_STRATEGY, "");
+
+static_assert(PAL_Z_DEFLATED == Z_DEFLATED, "");
```
Comment (stephentoub) [inline src/Native/System.IO.Compression.Native/pal_zlib.cpp]: Do we use all of these?  If not, we can trim away on both the managed and native sides the ones we don't.

File: src/Native/System.IO.Compression.Native/pal_zlib.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+/*
+Initializes the PAL_ZStream by creating and setting its underlying z_stream.
+*/
+static void Init(PAL_ZStream* stream)
+{
```
Comment (stephentoub) [inline src/Native/System.IO.Compression.Native/pal_zlib.cpp]:
```C++
assert(stream);
```

File: src/Native/System.IO.Compression.Native/pal_zlib.cpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    zStream->opaque = Z_NULL;
+
+    stream->internalState = zStream;
+}
+
+/*
+Frees any memory on the PAL_ZStream that was created by Init.
+*/
+static void End(PAL_ZStream* stream)
+{
```
Comment (stephentoub) [inline src/Native/System.IO.Compression.Native/pal_zlib.cpp]: ```assert(stream);```... I won't comment on these further, but we should do such asserts throughout.

Comment (stephentoub) [inline src/Native/System.IO.Compression.Native/pal_zlib.cpp]: Though I just noticed these two functions I commented on aren't extern.  We don't need to assert in static, just in the extern entrypoints to this shim.

File: src/Native/System.IO.Compression.Native/pal_zlib.cpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    z_stream* zStream = reinterpret_cast<z_stream*>(stream->internalState);
+
+    delete zStream;
```
Comment (stephentoub) [inline src/Native/System.IO.Compression.Native/pal_zlib.cpp]: We probably do want to assert here that zStream is not nullptr, as we're pulling it out of an object that's been passed around in managed code.

File: src/Native/System.IO.Compression.Native/pal_zlib.cpp [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+{
+    z_stream* zStream = reinterpret_cast<z_stream*>(stream->internalState);
+    TransferState(stream, zStream);
+    return zStream;
+}
+
+extern "C" int32_t IsZLibAvailable()
+{
+    return true;
+}
```
Comment (stephentoub) [inline src/Native/System.IO.Compression.Native/pal_zlib.cpp]: This doesn't seem right.  Previously we were P/Invoking into some function in libz, such that if we successfully found it, we knew zlib was available.  But now we're P/Invoking into the shim, whose presence doesn't mean zlib is installed.

File: src/Native/System.IO.Compression.Native/pal_zlib.cpp [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    
+    End(stream);
+
+    return result;
+}
+
+extern "C" int32_t IsCrc32Available()
+{
+    return true;
+}
```
Comment (stephentoub) [inline src/Native/System.IO.Compression.Native/pal_zlib.cpp]: This shouldn't need to be in the shim.  We had the check on the Windows side because clrcompression.dll is currently a trimmed down version of zlib, and we only just added the export for crc32, so this was a defensive mechanism to not try to use crc32 on the Windows side if it might not yet exist.  But if zlib is available on the Unix side, crc32 is going to be there as well.  I suggest just removing this from the shim, and if we need such a signature on the managed side to satisfy the managed PAL layer, we can just hardcode the true return value on the managed side.

File: src/System.IO.Compression/src/Interop/Interop.zlib.Windows.cs [hunk 7]
```diff
@@ -52,6 +52,23 @@ internal static partial class zlib
...
 
+        internal static unsafe bool IsCrc32Available()
+        {
+            try
+            {
+                // Make a P/Invoke into zlib crc32 to ensure we're able to find and use it.
+                // If we are, then use zlib.
+                crc32(0, null, 0);
+                return true;
+            }
```
Comment (stephentoub) [inline src/System.IO.Compression/src/Interop/Interop.zlib.Windows.cs]: @ianhays should comment, as I might be remembering incorrectly, but I seem to remember there being some issue recently with another project including this file and needing some code ifdef'd.

Comment (stephentoub) [inline src/System.IO.Compression/src/Interop/Interop.zlib.Windows.cs]: (After reading further, it looks like the ifdef is in another file, so maybe this is fine.)

File: src/System.IO.Compression/src/System/IO/Compression/DeflateStream.Unix.cs [hunk 8]
```diff
@@ -11,17 +11,18 @@ private static WorkerType GetDeflaterType()
...
-                // Make any P/Invoke into zlib to ensure we're able to find and use it.
-                // If we are, then use zlib.
-                Interop.zlib.zlibCompileFlags(); 
-                return WorkerType.ZLib;
+                if (Interop.zlib.IsZLibAvailable())
+                {
+                    return WorkerType.ZLib;
+                }
             }
             catch
```
Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/DeflateStream.Unix.cs]: If we're going to call into our own shim, we no longer need the try/catch, which was there to catch the P/Invoke exception that would arise from the export not being available.

File: src/System.IO.Compression/src/System/IO/Compression/ZLibNative.cs [hunk 9]
```diff
@@ -322,7 +315,7 @@ public ErrorCode DeflateInit2_(CompressionLevel level, int windowBits, int memLe
                 try { }
                 finally
                 {
-                    errC = Interop.zlib.DeflateInit2_(ref _zStream, level, CompressionMethod.Deflated, windowBits, memLevel, strategy, ZLibVersion);
+                    errC = Interop.zlib.DeflateInit2_(ref _zStream, level, CompressionMethod.Deflated, windowBits, memLevel, strategy);
```
Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/ZLibNative.cs]: We can remove the ```try { } finally { }``` and just leave the contents of the finally block, here and elsewhere this pattern is followed.

**Comment by stephentoub** [comment]
Thanks, Eric.  Generally looks good.  Nice to see the libraries file now only containing libraries associated with coreclr/corefx.

File: src/System.IO.Compression/src/Interop/Interop.zlib.Windows.cs (hunk 7 above)
Comment (eerhardt) [inline src/System.IO.Compression/src/Interop/Interop.zlib.Windows.cs]: I did a check in TFS, and the desktop .csproj file only includes Crc32Helper.cs (where it calls into this method), and not this Interop file.
I believe @ellismg was the one who included the #if to fix the build break in the desktop build. He can verify that we are safe here.

File: src/Native/System.IO.Compression.Native/pal_zlib.cpp (hunk 5 above)
Comment (eerhardt) [inline src/Native/System.IO.Compression.Native/pal_zlib.cpp]: I really struggled with this. So I decided to make it simple for now and have a conversation about what to do here.

My thinking was that if zlib isn't available on the platform, then we could do a #if HAVE_ZLIB return true; #else return false; here.  However, all the platforms we are supporting: Ubuntu, OSX, and CentOS all have it available. I'm not sure on FreeBSD yet.
So I was thinking to "rough the support in" like this.

If I was to take the shim IsZLibAvailable() out and return `WorkerType.ZLib` in DeflateStream.Unix.cs, then both Windows and Unix would always use the ZLib in/deflators, which means the Managed backup code would be dead code. So should that then be removed? I didn't want to go that far in this initial implementation.

I figured if we still wanted to support not having a zlib installed on the platform, this seemed like the most appropriate approach to enable that - a compile time check that would make this method return false. But since I don't know of a platform that doesn't have zlib, I wouldn't be able to test the compile time check, so I didn't add it (yet).

Another option would be to do a LoadLibrary call here instead.

So I think our first question to answer is: as of now, do we still want to have support for Unix systems that don't have zlib installed?

File: src/Native/System.IO.Compression.Native/pal_zlib.cpp (hunk 6 above)
Comment (eerhardt) [inline src/Native/System.IO.Compression.Native/pal_zlib.cpp]: OK - I will change it to return true from that managed side.

I thought that crc32 was only available in recent versions of zlib, and that we might hit a Unix platform that had an older version without crc32. I thought that was why it wasn't available on Windows - because we had compiled an older version. I didn't know it was because we trimmed down the zlib code we compiled for Windows.

File: src/System.IO.Compression/src/System/IO/Compression/ZLibNative.cs (hunk 9 above)
Comment (eerhardt) [inline src/System.IO.Compression/src/System/IO/Compression/ZLibNative.cs]: My plan was to do that in a separate PR. Is that OK?

Comment (stephentoub) [inline src/System.IO.Compression/src/System/IO/Compression/ZLibNative.cs]: Yup :smile:

File: src/System.IO.Compression/src/Interop/Interop.zlib.Windows.cs (hunk 7 above)
Comment (ianhays) [inline src/System.IO.Compression/src/Interop/Interop.zlib.Windows.cs]: This is fine. The IsCrc32Available function isn't used on Desktop since FEATURE_ZLIB_CRC32 is always false.

The whole system around IsCrc32Available and DeflateStream.*.cs WorkerType is wonkier than I'd like, but that's something to solve another day. I'd love to see us switch to an entirely native compression implementation i.e. removal of the managed implementation. Then we could avoid the WorkerType and IsCrc32Available stuff. The only situation where we use it is on a Unix system without Zlib, but in that case I'd rather we pull in a native zlib package from NuGet.

File: src/Native/System.IO.Compression.Native/pal_zlib.h [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+/*
+A structure that holds the input and output values for the zlib functions.
+*/
+struct PAL_ZStream
+{
+    uint8_t* nextIn; // next input byte
+    uint8_t* nextOut; // next output byte should be put there
+
+    uint32_t availIn; // number of bytes available at nextIn
+    uint32_t availOut; // remaining free space at nextOut
```
Comment (sokket) [inline src/Native/System.IO.Compression.Native/pal_zlib.h]: NIT: having the uint32_t members before the pointer members below adds padding (and therefore more memory to marshal)

Comment (eerhardt) [inline src/Native/System.IO.Compression.Native/pal_zlib.h]: I did move them around from my original order, since I was getting a compiler warning/error.

This was the first time I had to deal with padding. Is it good practice to put all the pointers together and then the integer types? Is there some sort of general ordering practice?

Comment (sokket) [inline src/Native/System.IO.Compression.Native/pal_zlib.h]: The best way to avoid internal padding and have proper alignment is to put the widest members first, starting with pointers. This ensures that the struct starts aligned and each member is packed. On second look, because you have two 32bit ints next to each other, you may get away with it here (on 63 and 32 bit systems since the pointers will be 64 bit and two 32 bit ints can be packed together to keep alignment). Feel free to won't fix since this should actually be a no-op on our current supported systems (64 and 32 bit) but just something to keep in mind :) sorry about that!

Comment (eerhardt) [inline src/Native/System.IO.Compression.Native/pal_zlib.h]: I will make the change since it is considered "best practice".

**Comment by sokket** [comment]
Other than the other comments and my super nit, LGTM

**Comment by eerhardt** [comment]
I have pushed a commit for PR feedback.

I want to do the following things in a separate PR:
1. Remove the empty try-finally pattern used.
2. Change the usage of (CompressionLevel)6 to use CompressionLevel.Default, since we are now using different versions of zlib - it would be more appropriate to use what the loaded version's default is, than using our tests against an older version.

File: src/Native/System.IO.Compression.Native/pal_zlib.h [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+Returns a PAL_ErrorCode indicating success or an error number on failure.
+*/
+extern "C" int32_t InflateEnd(PAL_ZStream* stream);
+
+/*
+Gets a value indicating whether the Crc32 function is available.
+
+Returns 1 if the Crc32 function is available, otherwise 0.
+*/
+extern "C" int32_t IsCrc32Available();
```
Comment (stephentoub) [inline src/Native/System.IO.Compression.Native/pal_zlib.h]: This should be removed, right?

File: src/Native/System.IO.Compression.Native/pal_zlib.h [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+{
+    PAL_Z_DEFLATED = 8
+};
+
+/*
+Gets a value indicating whether zlib is currently available.
+
+Returns 1 if zlib is available, otherwise 0.
+*/
+extern "C" int32_t IsZLibAvailable();
```
Comment (stephentoub) [inline src/Native/System.IO.Compression.Native/pal_zlib.h]: This should be removed, right?

**Comment by stephentoub** [comment]
Left a few more comments, but otherwise LGTM. Thanks.

File: src/Native/System.IO.Compression.Native/pal_zlib.h (hunk 11 above)
Comment (eerhardt) [inline src/Native/System.IO.Compression.Native/pal_zlib.h]: Yes, I missed these two. Thanks.

File: src/Native/System.IO.Compression.Native/pal_zlib.h (hunk 12 above)
Comment (eerhardt) [inline src/Native/System.IO.Compression.Native/pal_zlib.h]: Yes, I missed these two. Thanks.
